### PR TITLE
Add BV solver bitblast.

### PR DESCRIPTION
**Pull request by mpreiner**
This PR adds a new bit-blasting BV solver, which can be enabled via `--bv-solver=bitblast`. The new bit-blast solver can be used instead of the lazy BV solver and currently supports CaDiCaL and CryptoMiniSat as SAT back end.

### CODE REVIEW

**Comment by martin-cs** [comment]
👀❤️🎉🎉🎉

File: src/prop/cadical.h [hunk 1]
```diff
@@ -74,6 +75,7 @@ class CadicalSolver : public SatSolver
   void init();
 
   std::unique_ptr<CaDiCaL::Solver> d_solver;
+  std::vector<SatLiteral> d_assumptions;
```
Comment (ajreynol) [inline src/prop/cadical.h:78]: Can you add a one line doc on this?

File: src/prop/sat_solver.h [hunk 2]
```diff
@@ -96,6 +96,11 @@ class SatSolver {
   /** Check if the solver is in an inconsistent state */
   virtual bool ok() const = 0;
 
+  virtual void getUnsatAssumptions(std::vector<SatLiteral>& unsat_assumptions)
```
Comment (ajreynol) [inline src/prop/sat_solver.h:99]: Document: what does this correspond to and when are you allowed to call this.

File: src/theory/bv/bitblast/simple_bitblaster.cpp [hunk 3]
```diff
@@ -53,6 +53,10 @@ void BBSimple::storeBBTerm(TNode node, const Bits& bits)
 
 bool BBSimple::hasBBAtom(TNode atom) const
 {
+  if (atom.getKind() == kind::NOT)
```
Comment (ajreynol) [inline src/theory/bv/bitblast/simple_bitblaster.cpp:56]: Can you rename the argument to this method `lit`?  When I see `atom` as an argument this usually needs non-negated.

File: src/theory/bv/bitblast/simple_bitblaster.h [hunk 4]
```diff
@@ -58,6 +58,8 @@ class BBSimple : public TBitblaster<Node>
 
   prop::SatSolver* getSatSolver() override { Unreachable(); }
 
+  bool isVariable(TNode node);
```
Comment (ajreynol) [inline src/theory/bv/bitblast/simple_bitblaster.h:61]: Document.

File: src/theory/bv/bv_solver_bitblast.cpp [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    NodeManager* nm = NodeManager::currentNM();
+    d_inferManager.conflict(nm->mkAnd(conflict));
+  }
+}
+
+bool BVSolverBitblast::preNotifyFact(
+    TNode atom, bool pol, TNode fact, bool isPrereg, bool isInternal)
+{
+  d_facts.push_back(fact);
+  return false;
```
Comment (ajreynol) [inline src/theory/bv/bv_solver_bitblast.cpp:108]: Can you state why you return false?  (I believe this enables equality engine reasoning, right?)

**Review by ajreynol — CHANGES_REQUESTED** [review]

**Review by ajreynol — APPROVED** [review]
LGTM

**Comment by florianschanda** [comment]
This looks exciting :)

**Pull request merged**